File: leetcode/solutions/diameter-of-binary-tree/sol.cpp
// Problem: https://leetcode.com/problems/diameter-of-binary-tree/

// Recursive DFS

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int maxDepthModified(TreeNode* root, int* diameter) {
        // Returns the depth below the given node
        if (root == NULL) {
            // Diameter does not include the edge from leaf node to NULL,
            // so return -1 to adjust the overcalculation done earlier.
            return -1;
        }

        int left_depth = maxDepthModified(root -> left, diameter);
        int right_depth = maxDepthModified(root -> right, diameter);

        *diameter = max(*diameter, left_depth + right_depth + 2);

        return max(left_depth, right_depth) + 1;
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        // We can't use "the two deepest nodes" and add them up, because
        // could have some common ancestor

        int diameter = -1;
        maxDepthModified(root, &diameter);
        return diameter;
    }
};

File: leetcode/solutions/binary-tree-level-order-traversal/sol.py
from typing import List, Optional
from collections import deque

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []

        q = deque()
        q.append(root)
        ans = []

        while q:
            ans.append(list(x.val for x in q))

            for _ in range(len(q)):
                node = q.popleft()
                if node.left is not None:
                    q.append(node.left)

                if node.right is not None:
                    q.append(node.right)

        return ans

